core/src/main/java/io/seata/core/rpc/ChannelManager.java

```diff
@@ -118,54 +128,74 @@ private static RpcContext buildChannelHolder(NettyPoolKey.TransactionRole client
+            IDENTIFIED_CHANNELS.remove(expiredChannel);
+            LOGGER.info("remove expired tm client channel[{}] from identified channels.", expiredChannel);
+        }
+        TM_CLIENT_ACTIVE_CHANNELS.put(clientIdentified, newChannel);
+
         RpcContext rpcContext = buildChannelHolder(NettyPoolKey.TransactionRole.TMROLE, request.getVersion(),
             request.getApplicationId(),
             request.getTransactionServiceGroup(),
-            null, channel);
+            null, newChannel);
         rpcContext.holdInIdentifiedChannels(IDENTIFIED_CHANNELS);
-        String clientIdentified = rpcContext.getApplicationId() + Constants.CLIENT_ID_SPLIT_CHAR
-            + ChannelUtil.getClientIpFromChannel(channel);
-        TM_CHANNELS.putIfAbsent(clientIdentified, new ConcurrentHashMap<Integer, RpcContext>());
+
+        TM_CHANNELS.putIfAbsent(clientIdentified, new ConcurrentHashMap<>());
         ConcurrentMap<Integer, RpcContext> clientIdentifiedMap = TM_CHANNELS.get(clientIdentified);
         rpcContext.holdInClientChannels(clientIdentifiedMap);
     }
 
+    private static String getClientIdentified(String applicationId, Channel channel) {
+        return applicationId + Constants.CLIENT_ID_SPLIT_CHAR + ChannelUtil.getClientIpFromChannel(channel);
+    }
+
     /**
      * Register rm channel.
      *
-     * @param resourceManagerRequest the resource manager request
-     * @param channel                the channel
+     * @param request the resource manager request
+     * @param newChannel the new channel
      * @throws IncompatibleVersionException the incompatible  version exception
      */
-    public static void registerRMChannel(RegisterRMRequest resourceManagerRequest, Channel channel)
+    public static void registerRMChannel(RegisterRMRequest request, Channel newChannel)
         throws IncompatibleVersionException {
-        Version.checkVersion(resourceManagerRequest.getVersion());
-        Set<String> dbkeySet = dbKeytoSet(resourceManagerRequest.getResourceIds());
-        RpcContext rpcContext;
-        if (!IDENTIFIED_CHANNELS.containsKey(channel)) {
-            rpcContext = buildChannelHolder(NettyPoolKey.TransactionRole.RMROLE, resourceManagerRequest.getVersion(),
-                resourceManagerRequest.getApplicationId(), resourceManagerRequest.getTransactionServiceGroup(),
-                resourceManagerRequest.getResourceIds(), channel);
-            rpcContext.holdInIdentifiedChannels(IDENTIFIED_CHANNELS);
-        } else {
-            rpcContext = IDENTIFIED_CHANNELS.get(channel);
-            rpcContext.addResources(dbkeySet);
+        Version.checkVersion(request.getVersion());
+        Set<String> dbkeySet = dbKeytoSet(request.getResourceIds());
+
+        // remove channel from identified channels if expired.
+        String clientIdentified = getClientIdentified(request.getApplicationId(), newChannel);
+        Channel expiredChannel = RM_CLIENT_ACTIVE_CHANNELS.get(clientIdentified);
```

Comment:
If more than one client IP has the same applicationId application how to distinguish?